draft-irtf-cfrg-vdaf.md

```diff
@@ -3062,6 +3062,129 @@ measurement is sharded. This is provided to the FLP by Prio3.
+an integer in the range `[0, 2^bits)`. It is RECOMMENDED to set `chunk_length`
+to an integer near to the square root of `length * bits`. The optimal choice for
+any measurement size will vary slightly due to rounding.
+
+This instance uses PrgSha3 ({{prg-sha3}}) as its PRG. Its validity circuit,
+denoted `SumVec`, uses `Field128` ({{fields}}) as its finite field.
+
+Measurements are encoded as a vector of field elements with length
+`length * bits`. The field elements in the encoded vector represent all the bits
+of the measurement vector's elements, consecutively, in LSB to MSB order:
+
+~~~
+def encode(SumVec, measurement: Vec[Unsigned]):
+    if len(measurement) != SumVec.length:
+        raise ERR_INPUT
+
+    encoded = []
+    for elem in measurement:
+        if 0 > elem or elem >= 2 ** SumVec.bits:
+            raise ERR_INPUT
+
+        for l in range(SumVec.bits):
+            encoded.append(Sum.Field((elem >> l) & 1))
+
+    return encoded
+
+def truncate(SumVec, meas):
+    truncated = [SumVec.Field(0) for _ in range(SumVec.length)]
+    for i in range(SumVec.length):
+        for j in range(SumVec.bits):
+            weight = SumVec.Field(1 << j)
+            truncated[i] += weight * meas[i * SumVec.bits + j]
+    return truncated
+
+def decode(SumVec, output, _num_measurements):
+    return [x.as_unsigned() for x in output]
+~~~
+
+This validity circuit uses a `ParallelSum` gadget to achieve a smaller proof
+size. This optimization for "parallel-sum circuits" is described in
+{{BBCGGI19}}, section 4.4. Briefly, for circuits that add up the output of
+multiple identical sub-circuits, it is possible to achieve smaller proof sizes
+(on the order of O(sqrt(n)) instead of O(n)) by packaging more than one such
+subcircuit into a gadget.
+
+The `ParallelSum` gadget is parameterized with an arithmetic subcircuit, and a
+`count` of how many times it evaluates that subcircuit. It takes in a list of
+inputs and passes them through to instances of the subcircuit in the same order.
+It returns the sum of the subcircuit outputs.
+
+~~~
+def ParallelSum(self, inputs):
+    out = Field(0)
+    for i in range(self.count):
+        start_index = i * self.subcircuit.ARITY
+        end_index = (i + 1) * self.subcircuit.ARITY
+        out += self.subcircuit(inputs[start_index:end_index])
+    return out
+~~~
+
+The validity circuit checks that the encoded measurement consists of ones and
+zeros. Rather than use the `Range2` gadget on each element, it instead uses a
+`Mul` subcircuit and "free" constant multiplication and addition gates to
+simultaneously evaluate the same range check polynomial on each element, and
+multiply by a constant. (Arranging the circuit this way allows us to use the
+`ParallelSum` gadget, because the summands are identical subcircuits. Using
+`Range2` and then multiplying by different constants would not fit the parallel
+sum pattern.) One of the two `Mul` subcircuit inputs is equal to a measurement
```

Comment:
This parenthetical seems only useful if you're familiar with the (non-standardized) VDAF in libprio-rs with the same name (which we plan to deprecate when we cut the next draft). I'd suggest adding a bit more context here, perhaps describing in greater detail how the "naive" approach to parallelizing `Prio3Sum`would work (i.e., what we have in libprio-rs today). Alternatively, we might just cut this.

Reply:
My intent was to compare to the `Prio3Sum` circuit, and explain why it isn't in a parallel sum form. I think I'll cut the parenthetical.

Reply:
Ohhh gotcha, I misunderstood the intent. I think what I would do is have separate paragraph somewhere that is dedicated to contrasting Prio3Sum with Prio3SumVec, perhaps even with its own heading. Totally optional, but I think we're likely to get this question when the CFRG reviews this draft.